pkg/vm/vm.go

```diff
@@ -18,16 +18,16 @@ type VirtualMachineConfig struct {
 }
 
 type VirtualMachine struct {
-	Context              Context
-	MemorySegmentManager mem.MemorySegmentManager
-	Config               VirtualMachineConfig
+	Context       Context
+	MemoryManager mem.MemoryManager
+	Config        VirtualMachineConfig
 }
 
 // NewVirtualMachine creates a VM from the program bytecode using a specified config.
 func NewVirtualMachine(programBytecode *[]f.Element, config VirtualMachineConfig) *VirtualMachine {
 	return &VirtualMachine{
 		Context{Fp: 0, Ap: 0, Pc: 0},
```

Comment:
Shouldn't Fp, Ap, and Pc be MemoryValues now rather than uints?

Reply:
Should them? I am actually not sure that all three need to be MemoryValues. I think we can easily swap them to Memory Address if we require it.

Reply:
At least Fp and Ap should be, because for each function call a separate segment created for it's local memory and Fp and Ap are living in that segment during the execution of the func

Reply:
Pc is also a MemoryValue but in a dummy way, we know its segment is

Reply:
I am still not sure they need to be `MemoryValues` per se. Following the whitepaper recommendations on memory layout, function calls don't create new segments, only a single segment is used to store the stack trace which includes fp and pc.

`ap` is the allocation pointer, and will always point to free memory. We will always know its position, since it starts at 0, and every write operation increase it (assuming valid CASM). Then I believe that by induction we can prove we will always know its fixed location, since we know it is its previous location (which is fixed) + 1 , hence no need to use a `MemoryAddress`

Implementation wise we can see this as a segment where all the "value" types are stored as well as all the pointers to all reference types (this types getting segments of their own). This segment is going to get filled during a program execution at the same rate `ap` increases.

Since `fp` is just a snapshots of `ap` for everytime there is a function call,  we can say that for transitivity the same proof for `ap` applies to `fp`. So no need for `fp` to be a memory address as well.

Reply:
The old Python vm used to create a separate segment for each call, but ofc a vm is free to choose any method of segmenting it prefers.